### PR TITLE
Enhance TwoMaterialPropertyInterface and make InternalSideUserObject use it

### PR DESCRIPTION
**Pull request by YaqiWang**
Close issue #3961 and #5375.

### CODE REVIEW

**Comment by YaqiWang** [comment]
Can somebody review this? The newly added test fails in parallel. Any idea what I am missing?

**Comment by YaqiWang** [comment]
Well, a small test showed the problem is caused by stateful material...

**Comment by YaqiWang** [comment]
There is a bug in `ComputeMaterialsObjectThread`. I will update my PR shortly for fixing it.

**Comment by YaqiWang** [comment]
Ready for your review. I need this to fix one of my MR in Yak. Thanks!

**Comment by permcody** [comment]
@andrsd - can you review and merge this one?

File: framework/include/materials/TwoMaterialPropertyInterface.h [hunk 1]
```diff
@@ -26,19 +26,32 @@
 class TwoMaterialPropertyInterface : public MaterialPropertyInterface
 {
 public:
+  ///@{
+  /**
+   * Contructor.
+   *
+   * @param parameters The objects input parameters
+   * @param block_ids A reference to the block ids (optional)
+   * @param boundary_ids A reference to the boundary ids (optional)
```
Comment (andrsd) [inline framework/include/materials/TwoMaterialPropertyInterface.h]: `boundary_ids` is not this c-tor parameter

File: test/src/userobjects/InsideUserObject.C [hunk 2]
```diff
@@ -49,7 +49,7 @@ void
 InsideUserObject::execute()
 {
   for (unsigned int qp = 0; qp < _q_point.size(); ++qp)
-    _value += std::pow(_u[qp] - _u_neighbor[qp], 2) + (_diffusivity_prop ? (*_diffusivity_prop)[qp] : 0);
+    _value += std::pow(_u[qp] - _u_neighbor[qp], 2) + (_diffusivity_prop[qp] + _neighbor_diffusivity_prop[qp] )/2;
```
Comment (andrsd) [inline test/src/userobjects/InsideUserObject.C]: Add spaces around `/`.

File: test/tests/userobjects/internal_side_user_object/internal_side_user_object_two_materials.i [hunk 3]
```diff
@@ -0,0 +1,82 @@
+[Mesh]
+  type = FileMesh
+  file = TwoBlockMesh.e
```
Comment (andrsd) [inline test/tests/userobjects/internal_side_user_object/internal_side_user_object_two_materials.i]: I prefer `two_block_mesh.e` rather then `TwoBlockMesh.e`

File: framework/include/materials/TwoMaterialPropertyInterface.h (hunk 1 above)
Comment (YaqiWang) [inline framework/include/materials/TwoMaterialPropertyInterface.h]: yeah, copy-paste error...

File: test/tests/userobjects/internal_side_user_object/internal_side_user_object_two_materials.i (hunk 3 above)
Comment (YaqiWang) [inline test/tests/userobjects/internal_side_user_object/internal_side_user_object_two_materials.i]: Will change. BTW, if my other PR is accepted, we do not need this mesh, it is basically

```
nx=2
ny=2
subdomain_ids = '1 0
                             1 1'
```

Comment (andrsd) [inline test/tests/userobjects/internal_side_user_object/internal_side_user_object_two_materials.i]: If you want to wait for your other PR to be merge and then rebase this on top of it, that's fine with me.

**Comment by YaqiWang** [comment]
This PR now depends on my other PR with the newly-added test. So that PR needs to be merged to make all checks pass.

**Comment by YaqiWang** [comment]
I need this on-hold a little, because when I use this feature in yak, the address from material prop declaration is different from the address of calling getNeighborMaterialProperty.

**Comment by YaqiWang** [comment]
Turns out I made a bug in 4bee5dd, I did not use the deduced name for getting the property. Strangely, moose does not error out in this case.

**Comment by YaqiWang** [comment]
@andrsd This PR can not be retested and reviewed.

File: framework/include/materials/TwoMaterialPropertyInterface.h [hunk 4]
```diff
@@ -26,19 +26,31 @@
 class TwoMaterialPropertyInterface : public MaterialPropertyInterface
 {
 public:
+  ///@{
+  /**
+   * Contructor.
```
Comment (andrsd) [inline framework/include/materials/TwoMaterialPropertyInterface.h]: s/Contructor/Constructor/

File: framework/include/materials/TwoMaterialPropertyInterface.h [hunk 5]
```diff
@@ -26,19 +26,31 @@
...
+   *
+   * @param parameters The objects input parameters
+   * @param block_ids A reference to the block ids (optional)
+   *
+   * This class has four constructors:
```
Comment (andrsd) [inline framework/include/materials/TwoMaterialPropertyInterface.h]: Four c-tors? I see only 2.

File: framework/include/materials/TwoMaterialPropertyInterface.h [hunk 6]
```diff
@@ -26,19 +26,31 @@
 class TwoMaterialPropertyInterface : public MaterialPropertyInterface
 {
 public:
+  ///@{
```
Comment (andrsd) [inline framework/include/materials/TwoMaterialPropertyInterface.h]: Remove this line... Or add the closing doxy comment.

File: framework/include/materials/TwoMaterialPropertyInterface.h [hunk 7]
```diff
@@ -46,24 +58,48 @@ class TwoMaterialPropertyInterface : public MaterialPropertyInterface
 
 
 template<typename T>
-MaterialProperty<T> &
+const MaterialProperty<T> &
 TwoMaterialPropertyInterface::getNeighborMaterialProperty(const std::string & name)
 {
-  return _neighbor_material_data.getProperty<T>(name);
+  // Check if the supplied parameter is a valid imput parameter key
```
Comment (andrsd) [inline framework/include/materials/TwoMaterialPropertyInterface.h]: s/imput/input/

File: framework/include/materials/TwoMaterialPropertyInterface.h [hunk 8]
```diff
@@ -46,24 +58,48 @@ class TwoMaterialPropertyInterface : public MaterialPropertyInterface
...
+    return _neighbor_material_data.getProperty<T>(prop_name);
 }
 
 template<typename T>
-MaterialProperty<T> &
+const MaterialProperty<T> &
 TwoMaterialPropertyInterface::getNeighborMaterialPropertyOld(const std::string & name)
 {
-  return _neighbor_material_data.getPropertyOld<T>(name);
+  // Check if the supplied parameter is a valid imput parameter key
```
Comment (andrsd) [inline framework/include/materials/TwoMaterialPropertyInterface.h]: s/imput/input/

File: framework/include/materials/TwoMaterialPropertyInterface.h [hunk 9]
```diff
@@ -46,24 +58,48 @@ class TwoMaterialPropertyInterface : public MaterialPropertyInterface
...
+    return _neighbor_material_data.getPropertyOld<T>(prop_name);
 }
 
 template<typename T>
-MaterialProperty<T> &
+const MaterialProperty<T> &
 TwoMaterialPropertyInterface::getNeighborMaterialPropertyOlder(const std::string & name)
 {
-  return _neighbor_material_data.getPropertyOlder<T>(name);
+  // Check if the supplied parameter is a valid imput parameter key
```
Comment (andrsd) [inline framework/include/materials/TwoMaterialPropertyInterface.h]: s/imput/input/

File: framework/include/materials/TwoMaterialPropertyInterface.h (hunk 4 above)
Comment (YaqiWang) [inline framework/include/materials/TwoMaterialPropertyInterface.h]: copy-paste is devel...

**Comment by YaqiWang** [comment]
@andrsd Can you review this again?

**Pull request merged**